### PR TITLE
Add pending cancel subscriptions to dynamic segment [MAILPOET-4109]

### PR DESCRIPTION
**Pull request by websupporter**
Fixes [MAILPOET-4109]

When a subscriber cancels her subscription, we still need to consider this subscription active until the current period expires. The WooCommerce Subscriptions plugin uses for those subscriptions the post_status "wc-pending-cancel".

I have added an integration test, which tests the WooCommerceSubscription filter. Since there is currently no other setup for 3rd party plugins in integration tests, I follow the current model and create the necessary tables and data structures in the test itself.

How to test:
1. Go To WooCommerce > Subscriptions
2. Add Subscription to some of the existing subscribers, make sure to select Active status and to add the subscription product item when adding a subscription
3. Go To MailPoet > Lists and create a new segment that counts active subscriptions on site
4. Go back to Segments list and press ReCalculate Now button to see if your segment counts at least one subscriber
5. Go to WooCommerce > Subscriptions and press cancel for the active subscription
6. Make sure you see Pending Cancellation status for that subscription
7. Go back to Segments list and check that the user with the pending cancellation still exists in the dynamic segment calculation.. you will need to press ReCalculate Now button again

[MAILPOET-4109]: https://mailpoet.atlassian.net/browse/MAILPOET-4109?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    );
+
+
+    $sql = sprintf(
+      "SELECT DISTINCT wp_mailpoet_subscribers.id FROM wp_mailpoet_subscribers INNER JOIN wp_postmeta postmeta ON postmeta.meta_key = '_customer_user' AND wp_mailpoet_subscribers.wp_user_id=postmeta.meta_value INNER JOIN wp_posts posts ON postmeta.post_id = posts.id AND posts.post_type = 'shop_subscription' AND posts.post_status IN('wc-active', 'wc-pending-cancel') INNER JOIN wp_woocommerce_order_items items ON postmeta.post_id = items.order_id AND order_item_type = 'line_item' INNER JOIN wp_woocommerce_order_itemmeta itemmeta ON itemmeta.order_item_id=items.order_item_id AND itemmeta.meta_key='_product_id' WHERE itemmeta.meta_value IN (%s) GROUP BY wp_mailpoet_subscribers.id HAVING COUNT(wp_mailpoet_subscribers.id) = 2",
+      implode(',', $this->products)
+    );
+
+    global $wpdb;
+    $result = $wpdb->get_results($sql);
```
Comment (costasovo) [inline mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php:125]: I see that this doesn't create any data in the DB and also the $result variable is not used. Isn't this some debug leftover or so?

**Review by costasovo — COMMENTED** [review]
Great job and kudos for adding the test for the whole filter! 👍 

I found just a probably leftover in the test code. See my comment.

File: mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php (hunk 1 above)
Comment (websupporter) [reply mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php:125]: Debug leftover... Thx!

File: mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    $this->assertCount(1, $foundSubscribers);
+    $this->assertTrue(in_array($toBeFoundEmail, $foundSubscribers));
+    $this->tester->deleteWordPressUser($notToBeFoundEmail);
+    $this->tester->deleteWordPressUser($toBeFoundEmail);
+  }
+
+  private function getEmailsFromQueryBuilder(QueryBuilder $builder): array {
+    /**
+     * @var SubscribersRepository
+     */
```
[the remark this answers is not in the document]
Comment (websupporter) [reply mailpoet/tests/integration/Segments/DynamicSegments/Filters/WooCommerceSubscriptionTest.php:139]: Just double checked. No, the type is recognized. Not sure why I added this. Will remove it.

**Review by costasovo — APPROVED** [review]
👍

**Comment by rohankacharegs** [comment]
GlobalStep Verified

**Pull request merged**